tensorflow_probability/python/distributions/empirical_test.py

```diff
@@ -0,0 +1,694 @@
+      dist = empirical.Empirical(samples=input_ph)
+      self.assertAllClose(self.evaluate(dist.mean()), expected_mean)
+
+  def testMode(self):
+    samples = [
+        [0, 1, 1, 2],
+        [[0, 0, 1], [0, 1, 1], [2, 2, 2]],
+        [[[0, 0, 0, 0], [0, 0, 1, 1]],
+         [[1, 1, 1, 2], [0, 1, 2, 2]]]
+    ]
+    expected_modes = [
+        1,
+        [0, 1, 2],
+        [[0, 0], [1, 2]]
+    ]
+
+    for sample, expected_mode in zip(samples, expected_modes):
+      input_ = tf.convert_to_tensor(sample, dtype=np.int32)
+      input_ph = tf.placeholder_with_default(
+          input=input_,
+          shape=input_.shape if self.static_shape else None)
+      dist = empirical.Empirical(samples=input_ph)
+      self.assertAllClose(self.evaluate(dist.mode()), expected_mode)
+
+  def testVarianceAndStd(self):
+    samples = [
+        [0, 1, 1, 2],
+        [[1, 3], [2, 6]]
+    ]
+    expected_variances = [
+        0.5,
+        [1., 4.],
+    ]
+
+    for sample, expected_variance in zip(samples, expected_variances):
+      input_ = tf.convert_to_tensor(sample, dtype=np.float32)
+      input_ph = tf.placeholder_with_default(
+          input=input_,
+          shape=input_.shape if self.static_shape else None)
+      dist = empirical.Empirical(samples=input_ph)
+      self.assertAllClose(self.evaluate(dist.variance()),
+                          expected_variance)
+      self.assertAllClose(self.evaluate(dist.stddev()),
+                          np.sqrt(expected_variance))
+
+
+class EmpiricalVectorTest(test_util.VectorDistributionTestHelpers):
+
+  def testSamples(self):
+    for samples_shape in ([2, 4], [4, 2, 4], [2, 2, 2, 4]):
+      input_ = random_samples(samples_shape)
+      input_ph = tf.placeholder_with_default(
+          input=input_,
+          shape=samples_shape if self.static_shape else None)
+      dist = empirical.Empirical(samples=input_ph, event_ndims=1)
+      self.assertAllClose(input_ph, self.evaluate(dist.samples))
+
+    invalid_samples = [
+        0,
+        [0, 1],
+    ]
+    for samples in invalid_samples:
+      with self.assertRaises(ValueError):
+        dist = empirical.Empirical(samples=samples, event_ndims=1)
+
+  def testShapes(self):
+    for samples_shape in ([2, 4], [4, 2, 4]):
+      input_ = random_samples(samples_shape)
+      input_ph = tf.placeholder_with_default(
+          input=input_,
+          shape=samples_shape if self.static_shape else None)
+      dist = empirical.Empirical(samples=input_ph, event_ndims=1)
+      self.assertAllEqual(samples_shape[:-2],
+                          self.evaluate(dist.batch_shape_tensor()))
+      self.assertAllEqual(samples_shape[-1:],
+                          self.evaluate(dist.event_shape_tensor()))
+
+  def testCdfNoBatch(self):
+    sample = [[0, 0], [0, 1], [0, 1], [1, 2]]
+    events = [
+        [0, 1],
+        [[0, 1], [1, 2]],
+        [[[0, 0], [0, 1]],
+         [[1, 1], [1, 2]]]
+    ]
+    expected_cdfs = [
+        [0.75, 0.75],
```

Comment:
it looks to me like `events[0]` is a single event with batch `[]`, so i'm surprised to see `expected_cdfs[0]` isn't just `0.75`

then `events[1]` has batch shape `[2]`, so I would expect to see `expected_cdfs[1]` being `[0.75, 1.0]`

`events[2]` has batch shape `[2,2]`, so I expect `expected_cdfs[2]` to be `[[0.25, 0.75], [0.75, 1.0]]`

Reply:
Fixed.